pallets/foreign-investments/src/lib.rs

```diff
@@ -422,35 +380,673 @@ pub mod pallet {
+				if let Some(new_id) = new_swap_id {
+					SwapIdToForeignId::<T>::insert(new_id, (who.clone(), investment_id, action));
+					ForeignIdToSwapId::<T>::insert((who.clone(), investment_id, action), new_id);
+				}
+
+				if let Some(old_id) = previous_swap_id {
+					SwapIdToForeignId::<T>::remove(old_id);
+					ForeignIdToSwapId::<T>::remove((who.clone(), investment_id, action));
+				}
+			}
+
+			Ok(())
+		}
+
+		/// Apply a swap over a current possible swap state.
+		/// - If there was no previous swap, it adds it.
+		/// - If there was a swap in the same direction, it increments it.
+		/// - If there was a swap in the opposite direction:
+		///   - If the amount is smaller, it decrements it.
+		///   - If the amount is the same, it removes the inverse swap.
+		///   - If the amount is greater, it removes the inverse swap and create
+		///     another with the excess
+		///
+		/// The returned status contains the swapped amounts after this call and
+		/// the pending amounts to be swapped of both swap directions.
+		fn apply_swap(
+			who: &T::AccountId,
+			new_swap: SwapOf<T>,
+			over_swap_id: Option<T::SwapId>,
+		) -> Result<SwapStatus<T>, DispatchError> {
+			match over_swap_id {
+				None => {
+					let swap_id = T::TokenSwaps::place_order(
+						who.clone(),
+						new_swap.currency_in,
+						new_swap.currency_out,
+						new_swap.amount_in,
+						T::BalanceRatio::one(),
+					)?;
+
+					Ok(SwapStatus {
+						swapped: T::Balance::zero(),
+						pending: new_swap.amount_in,
+						swapped_inverse: T::Balance::zero(),
+						pending_inverse: T::Balance::zero(),
+						swap_id: Some(swap_id),
+					})
+				}
+				Some(swap_id) => {
+					let swap = T::TokenSwaps::get_order_details(swap_id)
+						.ok_or(Error::<T>::SwapOrderNotFound)?;
+
+					if swap.is_same_direction(&new_swap)? {
+						let amount_to_swap = swap.amount_in.ensure_add(new_swap.amount_in)?;
+						T::TokenSwaps::update_order(
+							who.clone(),
+							swap_id,
+							amount_to_swap,
+							T::BalanceRatio::one(),
+						)?;
+
+						Ok(SwapStatus {
+							swapped: T::Balance::zero(),
+							pending: amount_to_swap,
+							swapped_inverse: T::Balance::zero(),
+							pending_inverse: T::Balance::zero(),
+							swap_id: Some(swap_id),
+						})
+					} else {
+						let inverse_swap = swap;
+
+						let new_swap_amount_out = T::CurrencyConverter::stable_to_stable(
+							new_swap.currency_out,
+							new_swap.currency_in,
+							new_swap.amount_in,
+						)?;
+
+						match new_swap_amount_out.cmp(&inverse_swap.amount_in) {
+							Ordering::Less => {
+								let amount_to_swap =
+									inverse_swap.amount_in.ensure_sub(new_swap_amount_out)?;
+
+								T::TokenSwaps::update_order(
+									who.clone(),
+									swap_id,
+									amount_to_swap,
+									T::BalanceRatio::one(),
+								)?;
+
+								Ok(SwapStatus {
+									swapped: new_swap.amount_in,
+									pending: T::Balance::zero(),
+									swapped_inverse: new_swap.amount_in,
```

Comment:
Continuing with the discussion from #1692 (i.e. this [comment](https://github.com/centrifuge/centrifuge-chain/pull/1692/files/32d2297f731da83ff4d66e8c3e1ba60cd7ee6728#r1461601534)): The `pending` amount is definitely correct. However, I still believe we cannot set `swapped_inverse` amount to `new_swap.amount_in` because it is denoted in the opposite currency of `new_swap`, not the denomination of `inverse_swap`. 

So this should be `inverse_swap.amount_in.ensure_sub(amount_to_swap)` which equals `new_swap_amount_out`. This aligns with the `Ordering::Greater` implementation in which we set `swapped: inverse_swap_amount_out` (L563).

Reply:
You're right. I was paying attention to the wrong variable 🤦🏻‍♂️. Thanks for pushing this even though I disagree.